### PR TITLE
BFD-3073: support patient resource provider using HTTP POST

### PR DESCRIPTION
**Pull request by mackec1**
**JIRA Ticket:**
[BFD-3073](https://jira.cms.gov/browse/BFD-3073)

**User Story or Bug Summary:**
Supplement HTTP GET patient search with HTTP POST; GET query parameters will be encapsulated within a POST Body. This will limit BFD MDC logging from writing log entries where URL params might contain PII/PHI data.

By supporting query params within a POST body, we can safely pass in plain text (unhashed) MBI values to the Patient search operations.

### What Does This PR Do?
This PR represents continued work whereby BFD Patient/ services will support both HTTP GET and POST requests. In particular, support for passing in an unhashed MBI for a Patient lookup; thus for Patient/ lookup will now support:
- hashed HICN (V1 only)
- hashed MBI (V1 and V2)
- unhashed MBI (V1 and V2).
When dealing with unhashed MBI requests, only HTTP POST will be supported; any atttempt to pass in an unhashed MBI in an HTTP GET will trigger an InvalidRequestException. In addition the MDC logging code will further check Patient/ URL params to mask out any unhashed MBI if passed in as a URL param.

Normally, only bug fixes are addressed in V1 code, but this ticket represents work to safely support _Patient/_search_ in both V1 and V2; this is needed per RFC0020, as a means to remove necessity of hash algorithm sharing necessary under current HTTP GET URL params.

As part of the Patient search, some refactoring was implemented since V1 and V2 Patient provider classes had significant duplication of code; this was mitigated by introducing a CommonQueries.java which supported a more performant patient lookup, by introducing a database function, FIND_BENEFICIARY. The search for a unique BENE_ID supports either a hashed MBI, a hashed HICN, or an unhashed MBI. Adding to the types of queries also necessitate a change to the QueryLogListener class.

### What Should Reviewers Watch For?

* Does the implementation meet the criteria set forth in the AC.
* Were sufficient unit and integration tests implemented to vett the implementation.
* Does the implementation impose undue logging triggered by the QueryLogListener?
* Are there any unhandled and/or untested edge cases you can think of?
* Is user input properly sanitized & handled?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
* Changes to OpenAPI documentation will be addressed in: https://jira.cms.gov/browse/BFD-3092

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

* N/A

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by brandoncruz3 — DISMISSED** [review]

**Review by brandoncruz3 — DISMISSED** [review]
👍

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/RequestResponsePopulateMdcFilter.java [hunk 1]
```diff
@@ -49,6 +51,17 @@ public class RequestResponsePopulateMdcFilter extends OncePerRequestFilter {
...
+  /** obfuscated replacement for MBI. */
+  private static final String OBFUSCATED_MBI = "us-mbi|*********";
+
+  /** Regex Pattern to check for an /Patient request end-point. */
+  private static final Pattern PATIENT_REQUEST_REGEX = Pattern.compile("fhir/Patient");
+
+  /** Regex Pattern to check for an MBI. */
+  private static final Pattern MBI_REGEX =
+      Pattern.compile(
+          "us-mbi\\|[1-9][^SLOIBZsloibz0-9][^SLOIBZsloibz][0-9]-?[^SLOIBZsloibz0-9][^SLOIBZsloibz][0-9]-?[^SLOIBZsloibz0-9][^SLOIBZsloibz0-9][0-9][0-9]");
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/RequestResponsePopulateMdcFilter.java:63]: As far as I know, BFD's synthetic beneficiaries follow an MBI pattern with `/Ss/` in the second position, like our so-called golden beneficiary: `1S00ABBAA00`. Do you intend on excluding synthetic beneficiaries with this pattern?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    return (int) (values != null && values.size() > 0 ? values.get(0) : 0);
+  }
+
+  /**
+   * Search database for a BENE_ID using a combination of search type and search value. Supports use
+   * case where a beneficiary's MBI and/or HICN may have changed by looking at both current
+   * (beneficiaries) and historical (beneficiaries_history) data for a match.
+   *
+   * <p>For more information on the database function, see:
+   * V114__FIND_BENEFICIARY_IDENTIFIER_FUNCTION.SQL in the db migration directory.
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java:74]: Is this schema version and title accurate?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/RequestResponsePopulateMdcFilter.java (hunk 1 above)
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/RequestResponsePopulateMdcFilter.java:63]: I hadn't planned on it (unless you deem it necessary/useful). That 'filter' is obviously specific to a real MBI; I didn't see the necessity of filtering MBIs that weren't real.
As for testing this, I had to disable the HTTP POST assertion to even get access to a URL param...BUT, if it is a /Patient endpoint AND the us-mbi|<some real MBI> were to show up as a URL param, then it would be obfuscated.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+  /**
+   * Query database to determine which claim types have data for the specified beneficiary. The
+   * database function returns a bitwise mask value that denotes which claim types will have data
+   * for the specified beneficiaryId. This represents a fast and efficient way to ignore a requested
+   * claim type that ultimately has no data for our beneficiaryId.
+   *
+   * <p>The database function returns bits as follows: CARRIER_CLAIMS : bit 0 INPATIENT CLAIMS : bit
+   * 1 OUTPATIENT CLAIMS : bit 2 SNF CLAIMS : bit 3 DME CLAIMS : bit 4 HHA CLAIMS : bit 5 HOSPICE
+   * CLAIMS : bit 6 PART D CLAIMS : bit 7
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java:49]: I'm assuming the formatted mangled this. Could we avoid the formatter butchery with an unordered list?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    return (int) (values != null && values.size() > 0 ? values.get(0) : 0);
+  }
+
+  /**
+   * Search database for a BENE_ID using a combination of search type and search value. Supports use
+   * case where a beneficiary's MBI and/or HICN may have changed by looking at both current
+   * (beneficiaries) and historical (beneficiaries_history) data for a match.
+   *
+   * <p>For more information on the database function, see:
+   * V115__FIND_BENEFICIARY_IDENTIFIER_FUNCTION.SQL in the db migration directory.
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java:74]: This was recently updated to _V116__Find_Beneficiary_Identifier_Function.sql_, wasn't it?

Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java]: duh...don't know why I had 115 on brain...it is 116 and the sql is named correctly...thx.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/RequestResponsePopulateMdcFilter.java (hunk 1 above)
Comment (mjburling) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/RequestResponsePopulateMdcFilter.java:63]: Just to be clear, I was suggesting that the pattern **does** exclude the synthetic identifiers. As you said, that's obvious from the pattern, but my question was whether or not this was the intentional, and I think it is. I suppose my question is somewhat reactionary to the *very* recent events that had some folks confused about our synthetic beneficiary format. I believe the only place we reference documentation about MBI formatting is in the context of RFC0020.

No change necessary. I'm just working through how we can better document these things for ourselves and our users. Thanks!

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+   * <ul>
+   *   <li>CARRIER CLAIMS : bit 0
+   *   <li>INPATIENT CLAIMS : bit 1
+   *   <li>OUTPATIENT CLAIMS : bit 2
+   *   <li>SNF CLAIMS : bit 3
+   *   <li>DME CLAIMS : bit 4
+   *   <li>HHA CLAIMS : bit 5
+   *   <li>HOSPICE CLAIMS : bit 6
+   *   <li>PART D CLAIMS : bit 7
+   * </ul>
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java:58]: Thank you! No good deed goes unpunished, though...
```suggestion
   * <ul>
   *   <li>CARRIER CLAIMS : bit 0</li>
   *   <li>INPATIENT CLAIMS : bit 1</li>
   *   <li>OUTPATIENT CLAIMS : bit 2</li>
   *   <li>SNF CLAIMS : bit 3</li>
   *   <li>DME CLAIMS : bit 4</li>
   *   <li>HHA CLAIMS : bit 5</li>
   *   <li>HOSPICE CLAIMS : bit 6</li>
   *   <li>PART D CLAIMS : bit 7</li>
   * </ul>
```

Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java:58]: Yea, I wrestled with that a bit....
So other than CcwCodebookVariabe.java, we have oodles of 'not quite right HTML; try the following:
find . -name *.java |xargs grep "<li>"
I will change, but again I always seem to find myself having to change stuff that others don't.

Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java:58]: Added the '\<\/li\>'

`cmac:apps (cmac/BFD-3073-Support-Patient-Resource-Provider-POST) $ git push
Verifying Java source code format...
Verifying Java source code format for branch: cmac/BFD-3073-Support-Patient-Resource-Provider-POST ...
Relevant app-directory changes detected. Running fmt plugin check.
Inconsistencies discovered in formatting check. Run 'mvn com.spotify.fmt:fmt-maven-plugin:check --threads 1C' for details or 'mvn com.spotify.fmt:fmt-maven-plugin:format' to automatically apply the required formatting.`

Now barfs on formatting....so asked our formatter to fix formatting issues:
`mvn com.spotify.fmt:fmt-maven-plugin:format`

OK, now abides by our formatter....BUT
**Removed all the \</li\>**

**Review by mjburling — CHANGES_REQUESTED** [review]

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/QueryLoggingListener.java [hunk 6]
```diff
@@ -202,6 +202,8 @@ static enum QueryType {
     HIBERNATE_INFORMATION_SCHEMA_SEQUENCES(
         "hibernate_information_schema_sequences",
         (s -> s.contains(" from information_schema.sequences"))),
+    /** Represents query that invokes the check_claims_mask function. */
```
[the remark this answers is not in the document]
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/QueryLoggingListener.java]: fixed.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+      BfdMDC.put(
+          "database_query_by_hash_collision_distinct_bene_ids", Long.toString(values.size()));
+      throw new UnclassifiedServerFailureException(
+          404,
+          "By hash query found more than one distinct BENE_ID: "
+              + values.size()
+              + ", DistinctBeneIdsList: "
+              + values);
+    }
+    return values.get(0);
```
[the remark this answers is not in the document]
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java:152]: probably can't do that; `List<Long>` values is out of scope by the time we check to see what we can return.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/PatientE2EBase.java [hunk 8]
```diff
@@ -309,7 +393,7 @@ public void testPatientByIdentifierWhenSearchByHistoricalMbiExpectCurrentAndHist
     String historicalMbi = "HISTMBI";
     String expectedBeneTableMbi = "565654MBI";
     List<Object> loadedRecords = loadDataWithAdditionalBeneHistory();
-    // search mbi doesnt exist in the bene table, so bene history search must be true
+    // search mbi does'nt exist in the bene table, so bene history search must be true
```
[the remark this answers is not in the document]
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/PatientE2EBase.java]: fixed.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+   * @return int bitmask denoting which claims have data.
+   */
+  public static int availableClaimsData(EntityManager entityManager, long beneficiaryId) {
+    List<Object> values =
+        entityManager
+            .createNativeQuery(CHECK_CLAIMS_FOR_DATA_SQL)
+            .setParameter("beneIdValue", beneficiaryId)
+            .getResultList();
+
+    return (int) (values != null && values.size() > 0 ? values.get(0) : 0);
```
[the remark this answers is not in the document]
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java]: so I tried your suggestion and ITs started failing.
`
[ERROR] gov.cms.bfd.server.war.r4.providers.R4ExplanationOfBenefitResourceProviderTest.testFindByPatientSupportsWildcardClaimTypeV2 -- Time elapsed: 0.013 s <<< ERROR!
java.lang.NullPointerException: Cannot invoke "javax.persistence.Query.setParameter(String, Object)" because the return value of "javax.persistence.EntityManager.createNativeQuery(String, java.lang.Class)" is null
	at gov.cms.bfd.server.war.commons.CommonQueries.availableClaimsData(CommonQueries.java:71)
	at gov.cms.bfd.server.war.r4.providers.R4ExplanationOfBenefitResourceProvider.findByPatient(R4ExplanationOfBenefitResourceProvider.java:375)
	at gov.cms.bfd.server.war.r4.providers.R4ExplanationOfBenefitResourceProviderTest.testFindByPatientSupportsWildcardClaimTypeV2(R4ExplanationOfBenefitResourceProviderTest.java:511)
`
Suspect that the JDBC driver + Hibernate does not have a generic return type of List<Integer> but does support List<Object>. Now one can register new return types, but that is a pretty big chunk of code to gain 1-2 less lines of code.

**Review by brandoncruz3 — APPROVED** [review]

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mackec1) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CommonQueries.java:160]: also, List<String> rawValues = ... will cause rawValues to be null; the JDBC layer will return a ResultSet of Object(s)...JPA will not perform inline cast just because you specified a List<String>.

**Review by brandoncruz3 — APPROVED** [review]

**Review by rcrupp — DISMISSED** [review]
LGTM

**Review by rcrupp — APPROVED** [review]

**Review by aadedejifearless — APPROVED** [review]

**Review by brandoncruz3 — APPROVED** [review]
👍
